### PR TITLE
Implement Founders' Reward

### LINKED ISSUE #125: Implement the Founders' Reward.
For the Founders' Reward we desire these properties:

* the public can verify the total amount of the reward.
* the public can verify the rate at which the reward is disbursed / thaws. (see #96)
* the founders & investors are satisfied that the reward meets the agreed upon terms and distribution.

I also propose this engineering goal:

* the exception to consensus rules is minimal and contained in only the genesis block. eg: There is a single exceptional txOut containing all of the reward amount, and all subsequent distribution is through non-exceptional transactions.

A further goal might be:

* While the public can verify the reward total, the public cannot see the internal distribution. (Note that the founders/investors of course must be able to individually verify their reward amount.)

### PR DESCRIPTION
**Pull request by ebfull**
More info:  https://z.cash/blog/funding.html

The consensus rule is as follows:

```
All blocks before the first subsidy halving block, with the exception of
the genesis block, must contain an output which sends 20% of the block
subsidy value to a scriptPubKey `FOUNDERS_REWARD_SCRIPT`.
```

Right now, `FOUNDERS_REWARD_SCRIPT` is a 2-of-3 multisig P2SH.

Closes #125

### CODE REVIEW

**Comment by defuse** [comment]
ACK (but tests would be nice).

**Comment by ebfull** [comment]
Unfortunately, due to the impending changes with equihash and other situations, we'll have to ditch these tests.

**Comment by ebfull** [comment]
See #841

**Comment by nathan-at-least** [comment]
Reviewing…

File: src/main.cpp [hunk 1]
```diff
@@ -3004,6 +3004,22 @@ bool ContextualCheckBlock(const CBlock& block, CValidationState& state, CBlockIn
         }
     }
 
+    // Coinbase transaction must include founders' reward if the
+    // block height is below 210000, which is the first halvening.
```
Comment (nathan-at-least) [inline src/main.cpp]: Note the constant of 210000 blocks is correct now for the given block target rate.

However, we intend to change the block target time _and_ monetary base reward to be lower, then to reciprocally increase the halving intervals so that the curve looks like Bitcoin's with (roughly) the same time-since-launch scale.

File: src/miner.cpp [hunk 2]
```diff
@@ -104,13 +104,6 @@ CBlockTemplate* CreateNewBlock(const CScript& scriptPubKeyIn)
...
         pblock->nVersion = GetArg("-blockversion", pblock->nVersion);
 
-    // Create coinbase tx
-    CMutableTransaction txNew;
-    txNew.vin.resize(1);
-    txNew.vin[0].prevout.SetNull();
-    txNew.vout.resize(1);
-    txNew.vout[0].scriptPubKey = scriptPubKeyIn;
-
     // Add dummy coinbase tx as first transaction
```
Comment (nathan-at-least) [inline src/miner.cpp]: (Not a question on our diff, but rather upstream bitcoin:) Why is a dummy coinbase added rather than the actual coinbase?

Comment (ebfull) [inline src/miner.cpp]: What do you mean? If you're referring to an input, coinbase transactions always have one input with a null prevout to indicate they are coinbases.

File: src/miner.cpp [hunk 3]
```diff
@@ -329,8 +322,27 @@ CBlockTemplate* CreateNewBlock(const CScript& scriptPubKeyIn)
...
+        if (nHeight < FOUNDERS_REWARD_UNTIL_BLOCK) {
+            txNew.vout.resize(2);
+
+            auto rewardScript = ParseHex(FOUNDERS_REWARD_SCRIPT);
+
+            txNew.vout[0].scriptPubKey = CScript(rewardScript.begin(), rewardScript.end());
+            txNew.vout[0].nValue = FOUNDERS_REWARD_AMOUNT_PER_BLOCK * COIN;
+
+            txNew.vout[1].scriptPubKey = scriptPubKeyIn;
+            txNew.vout[1].nValue = (nFees + GetBlockSubsidy(nHeight, chainparams.GetConsensus())) - (FOUNDERS_REWARD_AMOUNT_PER_BLOCK * COIN);
```
Comment (nathan-at-least) [inline src/miner.cpp]: If either the definition of ``GetBlockSubsidy`` or ``consensusParams.nSubsidyHalvingInterval`` chanes and we don't also update ``FOUNDERS_REWARD_UNTIL_BLOCK``, this value may become incorrect for our intended monetary policy. Would it be less brittle to define the founders reward amount and block height cutoff inside ``ChainParams.cpp`` next to the other parameters?

**Comment by nathan-at-least** [comment]
NACK one comment detail (which is correct for this revision but I want to avoid giving people the wrong expectation if they latch on to constants which we're going to change).

ACK on all other code and concept.

Next up, I'm going to review buildbot logs and do some local build/testing.

File: src/miner.cpp (hunk 3 above)
Comment (ebfull) [inline src/miner.cpp]: I'm going to leave the constant there; we can refactor it out as we decide a number of other details, such as the mining slow start. (#762)

File: src/miner.cpp (hunk 2 above)
Comment (nathan-at-least) [inline src/miner.cpp]: Why does the old code say:

> CMutableTransaction txNew
> …
> pblock->vtx.push_back(CTransaction());

-instead of:

> CMutableTransaction txNew
> …
> pblock->vtx.push_back(txNew);

As this diff shows, ``txNew`` is not referenced until much later. Are those two different transactions?

**Comment by ebfull** [comment]
@zkbot try

**Comment by nathan-at-least** [comment]
We had some discussion on slack and I reviewed https://z.cash/blog/funding.html to recall these constraints:

a. halvings are "every four years" "just like Bitcoin"
b. Founder's Reward is defined to occur up to the first halvening
c. monetary base is defined in terms of approximate time: "40 coins per approximately 10 minutes to miners, 10 per approx 10 minutes to founders."

Let's replace the comment with just a link to the blogpost.

File: src/main.cpp [hunk 4]
```diff
@@ -3004,9 +3004,9 @@ bool ContextualCheckBlock(const CBlock& block, CValidationState& state, CBlockIn
         }
     }
 
-    // Coinbase transaction must include founders' reward if the
-    // block height is below 210000, which is the first halvening.
-    // The genesis block is excused from this requirement.
+    // Coinbase transaction must include founders' reward for the
+    // first four years. The genesis block is excused from this
+    // requirement.
```
Comment (nathan-at-least) [inline src/main.cpp]: One more request: Let's say "see https://z.cash/blog/funding.html".

File: src/main.cpp [hunk 5]
```diff
@@ -3004,6 +3004,22 @@ bool ContextualCheckBlock(const CBlock& block, CValidationState& state, CBlockIn
         }
     }
 
+    // Coinbase transaction must include founders' reward if the
+    // block height is below 210000, which is the first halvening.
+    // The genesis block is excused from this requirement.
+    if ((nHeight > 0) && (nHeight < FOUNDERS_REWARD_UNTIL_BLOCK)) {
```
Comment (daira) [inline src/main.cpp]: This should probably be ``&& (nHeight <= FOUNDERS_REWARD_UNTIL_BLOCK)``, since that makes the total reward to the founders equal to ``FOUNDERS_REWARD_UNTIL_BLOCK * FOUNDERS_REWARD_AMOUNT_PER_BLOCK`` (and perhaps it should be renamed to ``FOUNDERS_REWARD_BLOCKS``).

File: src/miner.cpp [hunk 6]
```diff
@@ -329,8 +322,27 @@ CBlockTemplate* CreateNewBlock(const CScript& scriptPubKeyIn)
         LogPrintf("CreateNewBlock(): total size %u\n", nBlockSize);
 
         // Compute final coinbase transaction.
-        txNew.vout[0].nValue = nFees + GetBlockSubsidy(nHeight, chainparams.GetConsensus());
+        CMutableTransaction txNew;
+        txNew.vin.resize(1);
+        txNew.vin[0].prevout.SetNull();
         txNew.vin[0].scriptSig = CScript() << nHeight << OP_0;
+
+        if (nHeight < FOUNDERS_REWARD_UNTIL_BLOCK) {
```
Comment (daira) [inline src/miner.cpp]: This should be the same condition as https://github.com/zcash/zcash/pull/831/commits/11483859b56142b391d4fa89d84a25e51d073568#diff-7ec3c68a81efff79b6ca22ac1f1eabbaR3010 (possibly it doesn't matter if we don't use this code to mine the genesis block, but it's confusing).

File: src/miner.cpp (hunk 3 above)
Comment (daira) [inline src/miner.cpp]: In any case I suggest making it
```
txNew.vout[1].nValue = nFees + GetBlockSubsidy(nHeight, chainparams.GetConsensus()) - txNew.vout[0].nValue;
```

Comment (daira) [inline src/miner.cpp]: Also you could write ``CAmount nSubsidy = GetBlockSubsidy(nHeight, chainparams.GetConsensus());`` outside the ``if`` to further reduce code duplication between the branches.

File: src/miner.cpp (hunk 6 above)
Comment (daira) [inline src/miner.cpp]: Oh I see you fixed this in the next commit.

**Comment by daira** [comment]
ACK modulo my comments and the lack of tests.

File: src/main.cpp (hunk 5 above)
Comment (ebfull) [inline src/main.cpp]: That won't be the case anyway because of the genesis block, but I want to avoid being inclusive here since we want to turn the founders reward off *before* the rewards halve.

File: src/miner.cpp (hunk 2 above)
Comment (ebfull) [inline src/miner.cpp]: The old code (and the new code) is just reserving a placeholder transaction for the coinbase. It isn't fully constructed yet, and must be replaced later, so it is irrelevant where it is constructed. (It *should* probably be constructed in the scope that I move it to.)

I think your question was worth asking though.

**Comment by ebfull** [comment]
Rebased!

File: src/main.cpp [hunk 7]
```diff
@@ -3010,6 +3010,26 @@ bool ContextualCheckBlock(const CBlock& block, CValidationState& state, CBlockIn
         }
     }
 
+    // Coinbase transaction must include an output sending 20% of
+    // the block reward to `FOUNDERS_REWARD_SCRIPT` until the first
+    // subsidy halving block, with exception to the genesis block.
+    if ((nHeight > 0) && (nHeight < consensusParams.nSubsidyHalvingInterval)) {
```
Comment (nathan-at-least) [inline src/main.cpp]: These bounds exclude the genesis block (to be consistent with upstream) and the first halved block, so that's (currently) 210000 - 1 blocks * 10 ZEC / block == 2099990 ZEC total as Founder's Reward. We're not certain if that's acceptable based on business requirements, and we're proceeding with this for now.

File: qa/rpc-tests/txn_doublespend.py [hunk 8]
```diff
@@ -24,39 +24,37 @@ def setup_network(self):
...
             assert_equal(self.nodes[i].getbalance(), starting_balance)
             self.nodes[i].getnewaddress("")  # bug workaround, coins generated assigned to first getnewaddress!
         
         # Assign coins to foo and bar accounts:
-        amtfoo = 970
-        amtbar = starting_balance - amtfoo
-        self.nodes[0].move("", "foo", amtfoo)
-        self.nodes[0].move("", "bar", amtbar)
+        self.nodes[0].move("", "foo", (starting_balance - (mining_reward - 20)))
+        self.nodes[0].move("", "bar", (mining_reward - 20))
```
Comment (nathan-at-least) [inline qa/rpc-tests/txn_doublespend.py]: +1, this is much clearer than my attempt on the left.

File: qa/rpc-tests/zcpour.py [hunk 9]
```diff
@@ -13,91 +13,38 @@
...
 
-        (total_in, inputs) = gather_inputs(self.nodes[1], 40)
-        protect_tx = self.nodes[1].createrawtransaction(inputs, {})
-        pour_result = self.nodes[1].zcrawpour(protect_tx, {}, {zcaddress:49.9}, 40, 0.1)
+        (total_in, inputs) = gather_inputs(self.nodes[0], 40)
+        protect_tx = self.nodes[0].createrawtransaction(inputs, {})
+        pour_result = self.nodes[0].zcrawpour(protect_tx, {}, {zcaddress:39.9}, 40, 0.1)
 
-        receive_result = self.nodes[1].zcrawreceive(zcsecretkey, pour_result["encryptedbucket1"])
+        receive_result = self.nodes[0].zcrawreceive(zcsecretkey, pour_result["encryptedbucket1"])
```
Comment (nathan-at-least) [inline qa/rpc-tests/zcpour.py]: Why did this node index change?

File: qa/rpc-tests/zcpour.py [hunk 10]
```diff
@@ -13,91 +13,38 @@
...
-        pour_result = self.nodes[1].zcrawpour(pour_tx, {receive_result["bucket"] : zcsecretkey}, {zcaddress: 49.8}, 0, 0.1)
-
-        self.nodes[1].sendrawtransaction(pour_result["rawtxn"])
-        self.nodes[1].generate(1)
+        pour_tx = self.nodes[0].createrawtransaction([], {})
+        pour_result = self.nodes[0].zcrawpour(pour_tx, {receive_result["bucket"] : zcsecretkey}, {zcaddress: 39.8}, 0, 0.1)
 
-        print "Syncing blocks..."
-        connect_nodes(self.nodes[1], 0)
-        sync_blocks(self.nodes[0:2])
```
Comment (nathan-at-least) [inline qa/rpc-tests/zcpour.py]: Why are these lines removed?

File: qa/rpc-tests/zcpour.py [hunk 11]
```diff
@@ -13,91 +13,38 @@
...
-
-        signed_tx_pour = self.nodes[0].signrawtransaction(pour_result["rawtxn"])
-
-        self.nodes[0].sendrawtransaction(signed_tx_pour["hex"])
-        
-        self.send_pours_around()
-
 if __name__ == '__main__':
-    PourTxTest().main()
+    PourTxTest().main()
```
Comment (nathan-at-least) [inline qa/rpc-tests/zcpour.py]: nitpick (non-blocking): New newline at end of file. Let's always end with a newline.

**Comment by nathan-at-least** [comment]
ACK. (Curious how the network magic numbers were selected, though.)

@zkbot r+

File: qa/rpc-tests/zcpour.py (hunk 9 above)
Comment (ebfull) [inline qa/rpc-tests/zcpour.py]: The test created a network before, I removed it so we only have one node.

File: qa/rpc-tests/zcpour.py (hunk 10 above)
Comment (ebfull) [inline qa/rpc-tests/zcpour.py]: There's no longer a network for nodes to sync with; it was pointless for this test.
